Add tooltip to show levels required for next combat level

bmiddle opened the pull request:
# What I've added
* New VarClientInt for the interface's current tab called `CURRENT_TAB`
* Created a tooltip that calculates the levels required to increase your combat level
* New configuration setting, calculating next combat level is defaulted to false

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/combatlevel/CombatLevelPlugin.java:
```diff
@@ -89,4 +111,81 @@ public void updateCombatLevel(GameTick event)
 		combatLevelWidget.setText("Combat Lvl: " + decimalFormat.format(combatLevelPrecise));
 	}
+
+	Map<Skill, Integer> getLevelsUntil()
```
Why you put this into plugin when only place where you use it is overlay?

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/combatlevel/CombatLevelOverlay.java:
```diff
@@ -0,0 +1,58 @@
+package net.runelite.client.plugins.combatlevel;
```
Add license header

bmiddle replied:
Added both license headers

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/combatlevel/CombatLevelConfig.java:
```diff
@@ -0,0 +1,23 @@
+package net.runelite.client.plugins.combatlevel;
```
Add license header

Kamielvf commented on runelite-api/src/main/java/net/runelite/api/VarClientInt.java:
```diff
@@ -34,7 +34,15 @@
+
+	/**
+	 * Current opened tab
+	 * 	Attack Styles = 0, Stats = 1, Quest log = 2, Inventory = 3, Worn Equipment = 4,
```
`-1` is side-panels closed

bmiddle commented:
Finished those requested changes, deathbeam

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/combatlevel/CombatLevelOverlay.java:
```diff
@@ -0,0 +1,169 @@
+		}
+	}
+
+	private int calcLevels(double start, int end, double multiple)
```
This can be static

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/combatlevel/CombatLevelOverlay.java:
```diff
@@ -0,0 +1,169 @@
+		return need;
+	}
+
+	private int calcLevelsRM(double start, int end, double dhp)
```
This can be static too

bmiddle commented:
Can this get another review? Travis failed on a caching test

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/combatlevel/CombatLevelOverlay.java:
```diff
@@ -0,0 +1,169 @@
+		return null;
+	}
+
+	private void getLevelsUntil()
```
you could probably just merge this with render method/return string builder instead of using map and then iterating through, and at places where you put to map instead append to stringbuilder

bmiddle commented:
I changed `getLevelsUntil()` to return a string which turned out to be a lot cleaner for the rendering method.

Adam- commented on runelite-api/src/main/java/net/runelite/api/VarClientInt.java:
```diff
@@ -34,7 +34,15 @@
+	 * 	Worn Equipment = 4, Prayer = 5, Magic = 6, Clan Chat = 7, Friends List = 8,
+	 *	Ignore List = 9, Logout = 10, Options = 11, Emotes = 12, Music Player = 13
+	 */
+	CURRENT_TAB(171);
```
It may be worth an enum for these values.

Adam- commented on runelite-client/src/main/java/net/runelite/client/plugins/combatlevel/CombatLevelOverlay.java:
```diff
@@ -0,0 +1,147 @@
+		}
+
+		Widget combatLevelWidget = client.getWidget(WidgetInfo.COMBAT_LEVEL);
+		Rectangle combatCanvas = combatLevelWidget.getBounds();
```
null check both of these.

Adam- commented on runelite-client/src/main/java/net/runelite/client/plugins/combatlevel/CombatLevelOverlay.java:
```diff
@@ -0,0 +1,147 @@
+			start += multiple;
+			need++;
+		}
+		return need;
```
is this not `start + ((end - start) / multiple)`?

bmiddle replied:
Would this be the return value?

bmiddle replied:
It works fine using `Math.ceil((end - start) / multiple)`

bmiddle commented:
@Adam-, I removed the loops and added much faster calculations for the `calcLevels(...)` and `calcLevelsRM(...)` methods, added the null checks (though I need someone to double check if I need the empty lines), and created a new enum for the indices of the interface's opened tab called `TabIndex`. This should be ready for another review.

bmiddle commented:
I noticed the prayer calculation was inaccurate, but I added a conditional to make it accurate before the tooltip gets created.

bmiddle commented:
Need an opinion, should this feature be defaulted on? Currently it's off, but I wasn't sure if people wanted to have it on automatically.

Jasper-ketelaar commented:
I would say having it on by default is fine.

bmiddle commented:
Can this be looked into again? Does it need rebasing, etc.?

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/combatlevel/CombatLevelOverlay.java:
```diff
@@ -0,0 +1,151 @@
+	@Override
+	public Dimension render(Graphics2D graphics)
+	{
+		if (!config.showLevelsUntil() || client.getLocalPlayer().getCombatLevel() == 126
```
I feel like defining 126 as the `MAXIMUM_COMBAT_LEVEL` in another API file would be useful. Perhaps in `Experience`?

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/combatlevel/CombatLevelOverlay.java:
```diff
@@ -0,0 +1,151 @@
+
+		// create and populate tooltip string
+		StringBuilder sb = new StringBuilder();
+		sb.append("<col=ff981f>Next combat level:</col></br>");
```
Use `ColorUtil`.

Nightfirecat commented:
Also, `calcLevels()` and `calcLevelsRM()` seem fairly dense... Javadoc or more descriptive variable names there might help that.

ahampe commented on runelite-client/src/main/java/net/runelite/client/plugins/combatlevel/CombatLevelOverlay.java:
```diff
@@ -0,0 +1,151 @@
+
+		if (prayerLevel % 2 == 0)
+		{
+			prayNeed++;
```
Just wanted to point out, I've noticed that https://oldschool.tools/calculators/combat-level also performs this check, and it incorrectly overestimates the number of prayer levels needed. I think the correct adjustment is `if ((prayerLevel + prayNeed) % 2 == 1) prayNeed--;`. Raising prayer to an odd level will never increase combat level.

ahampe commented on runelite-client/src/main/java/net/runelite/client/plugins/combatlevel/CombatLevelOverlay.java:
```diff
@@ -0,0 +1,151 @@
+		// begin calculating levels
+		int next = client.getLocalPlayer().getCombatLevel() + 1;
+		int meleeNeed = calcLevels(base + melee, next, 0.325);
+		int hpdefNeed = calcLevels(base + max, next, 0.25025);
```
Is "0.25025" correct? I thought it was just 0.25...

Nightfirecat commented:
Also, I would strongly recommend this feature includes tests (based on actual account combat levels if possible)

ShaunDreclin commented:
This will resolve #448
